oap-server/server-receiver-plugin/aws-firehose-receiver/src/main/java/org/apache/skywalking/oap/server/receiver/aws/firehose/AWSFirehoseReceiverModuleProvider.java

```diff
@@ -0,0 +1,99 @@
+
+public class AWSFirehoseReceiverModuleProvider extends ModuleProvider {
+    public static final String NAME = "default";
+
+    private AWSFirehoseReceiverModuleConfig moduleConfig;
+    private HTTPServer httpServer;
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public Class<? extends ModuleDefine> module() {
+        return AWSFirehoseReceiverModule.class;
+    }
+
+    @Override
+    public ConfigCreator<AWSFirehoseReceiverModuleConfig> newConfigCreator() {
+        return new ConfigCreator<AWSFirehoseReceiverModuleConfig>() {
+            @Override
+            public Class<AWSFirehoseReceiverModuleConfig> type() {
+                return AWSFirehoseReceiverModuleConfig.class;
+            }
+
+            @Override
+            public void onInitialized(final AWSFirehoseReceiverModuleConfig initialized) {
+                moduleConfig = initialized;
+            }
+        };
+    }
+
+    @Override
+    public void prepare() throws ServiceNotProvidedException {
+        final HTTPServerConfig httpServerConfig = HTTPServerConfig.builder().host(moduleConfig.getHost())
+                                                                  .port(moduleConfig.getPort())
+                                                                  .contextPath(moduleConfig.getContextPath())
+                                                                  .maxThreads(moduleConfig.getMaxThreads())
+                                                                  .idleTimeOut(moduleConfig.getIdleTimeOut())
+                                                                  .acceptQueueSize(moduleConfig.getAcceptQueueSize())
+                                                                  .maxRequestHeaderSize(
+                                                                      moduleConfig.getMaxRequestHeaderSize())
+                                                                  .enableTLS(moduleConfig.isEnableTLS())
+                                                                  .tlsKeyPath(moduleConfig.getTlsKeyPath())
+                                                                  .tlsCertChainPath(moduleConfig.getTlsCertChainPath())
+                                                                  .build();
+        httpServer = new HTTPServer(httpServerConfig);
+        httpServer.initialize();
+    }
+
+    @Override
+    public void start() throws ServiceNotProvidedException, ModuleStartException {
+        httpServer.addHandler(
+            new FirehoseHTTPHandler(),
+            Collections.singletonList(HttpMethod.POST)
+        );
+    }
+
+    @Override
+    public void notifyAfterCompleted() throws ServiceNotProvidedException, ModuleStartException {
+        httpServer.start();
+    }
+
+    @Override
+    public String[] requiredModules() {
+        return new String[] {
+            OtelMetricReceiverModule.NAME
```

Comment:
As you have decleared dependencies, please expose the otel processor as a service, and use in awsreceiver.

Reply:
There is no chance to inject an instance of `OpenTelemetryMetricRequestProcessor` to `OpenTelemetryMetricHandler`, as `OpenTelemetryMetricHandler` is initialized by reflection, so I use singleton.

In order to expose `OpenTelemetryMetricRequestProcessor` as service, I consider changing `OpenTelemetryMetricHandler` as interface, such as
```java
public interface OpenTelemetryMetricRequestProcessor extends Service {
    void processMetricsRequest(final ExportMetricsServiceRequest requests);
}
```
`OpenTelemetryMetricHandler` will implement `OpenTelemetryMetricRequestProcessor`:
```
public class OpenTelemetryMetricHandler
    extends MetricsServiceGrpc.MetricsServiceImplBase
    implements Handler, OpenTelemetryMetricRequestProcessor{
//.....
}
```
otel-module will register instance of `OpenTelemetryMetricHandler` as `OpenTelemetryMetricRequestProcessor` service on prepare stage, such as : 

```java

    @Override
    public void prepare() throws ServiceNotProvidedException, ModuleStartException {
        if (config.getEnabledHandlers().isEmpty()) {
            return;
        }
        this.handlers =
            Handler.all().stream()
                   .filter(h -> config.getEnabledHandlers().contains(h.type()))
                   .peek(h -> {
                       if (h instanceof OpenTelemetryMetricRequestProcessor) {
                           registerServiceImplementation(
                               OpenTelemetryMetricRequestProcessor.class,
                               (OpenTelemetryMetricRequestProcessor) h
                           );
                       }
                   }).collect(toList());

    }
```

That means otel-module doesn't  directly expose `OpenTelemetryMetricHandler`, `OpenTelemetryMetricRequestProcessor` instead. WDYT?

Reply:
I don't mean HTTP or gRPC handler as service, we should not do that.
The OpenTelemetryMetricRequestProcessor should hold all shared logic and variables as a service. Both handlers is initialized by it. What block you?

Reply:
Actually. I considered `OpenTelemetryMetricHandler` and `FirehoseHTTPHandler` hold `OpenTelemetryMetricRequestProcessor `, so I couldn't find appropriate chance to initialize `OpenTelemetryMetricRequestProcessor` and assign instance of `OpenTelemetryMetricRequestProcessor` to `OpenTelemetryMetricHandler`, as it's initialized by reflection.

>  Both handlers is initialized by it

Did you mean `OpenTelemetryMetricRequestProcessor` initialize `OpenTelemetryMetricHandler` and  `FirehoseHTTPHandler`?

Reply:
I don't have a laptop around me. I still can't understand why the data processor would block a handler. Handler is for rpc listening. Processor is for processing data which could be initialized and set to handler sooner or later. 
Why the boundaries seem not clear now?

Reply:
I try to read codes of OpenTelemetryMetricHandler. Generally it does 2 things
- Load MAL files and build the converter.
- Build gRPC relative handler for proto and rpc relative things.

For the new processor, We just need MAL relative part, and open an API for converters.

Reply:
Yep,  processor for MAL, and handler for RPC, it's clear. But there are two things that confuse me:

1. `converters` depends on  `MeterSystem`, which means `processorService` also need to do so. Registering a service must be on prepare stage, but  I only could get `MeterSystem` on start stage.  It has conflict.
2. How to find which handler requires processor and  inject processor into it in otel-receiver module?  How about we  add a method such as `setProcessor( processor)` to   `OpenTelemetryMetricHandler` , and if the handler's type is `OpenTelemetryMetricHandler`, we call the method to inject processor

Reply:
OK, I think you are limited by Zhenxu code style rather than OAP kernel.

https://github.com/apache/skywalking/blob/e497efd2450930c8a572d19d9ead4bbd9e08902a/oap-server/server-receiver-plugin/otel-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/otel/Handler.java#L33

Hanlder.all is using class scan to get all handler, but actually we don have to.
OC and otlp handler could be hard coded in provider, and than we have clear handler references about how to initial them.

Reply:
@kezhenxu94 I want to suggest to remove handler scan and separate otlp processor out of handler, which would be reused by awsreceiver. Do you have any concern for this?

@pg-yang I think we need a new PR for this to make sure review and change log is clear.

Reply:
> we add a method such as setProcessor( processor) to OpenTelemetryMetricHandler

This works, but I don't like instanceof, as this is a barrier we set for ourselves(class scan caused this). I pefer to init them(handlers) one by one in the provider.

Reply:
> @kezhenxu94 I want to suggest to remove handler scan and separate otlp processor out of handler, which would be reused by awsreceiver. Do you have any concern for this?
> 
> @pg-yang I think we need a new PR for this to make sure review and change log is clear.

I think scanning handler classes was initially wrote by @hanahmily and I just commented from the perspective of breaking changes here https://github.com/apache/skywalking/pull/10300#discussion_r1083250217

Reply:
Long history. 😂

Reply:
Scanning tends to get rid of "hard code". It's fine to remove it if it blocks current works. Nothing won't been broken.